core/xml.js

```diff
@@ -462,9 +407,178 @@ Blockly.Xml.domToWorkspace = function(xml, workspace) {
     workspace.setResizesEnabled(true);
   }
   Blockly.Events.fire(new Blockly.Events.FinishedLoading(workspace));
-  return newBlockIds;
+  return state.newBlockIds;
 };
 
+/**
+ * Defines the behavior for how the parser should handle errors. Projects using
+ * Blockly can override this method to provide project-specific error handling
+ * or to present more meaningful information to users. The default behavior is
+ * simply to throw the error.
+ * @param {!string} message The error message describing what went wrong.
+ * @public
+ */
+Blockly.Xml.onError = function(message) {
+  throw Error(message);
+};
+
+/**
+ * Defines the behavior for how the parser should handle warnings. Projects
+ * using Blockly can override this method to provide project-specific warning
+ * handling or to present more meaningful information to users. The default
+ * behavior is to log the message to the browser console.
+ * @param {!string} message The warning message describing what went wrong.
+ * @public
+ */
+Blockly.Xml.onWarning = function(message) {
+  console.warn(message);
+};
+
+/**
+ * A mapping of XML tags to parser functions.
+ * @type {Object.<string,
+ *     function(Blockly.Workspace, Element, Blockly.Xml.ParserState)>}
+ */
+Blockly.Xml.tagMap_ = {};
+
+/**
+ * A list of operations to serialize a workspace to XML.
+ * @type {Array.<function(Element, Blockly.Workspace, boolean)>}
+ * @private
+ */
+Blockly.Xml.serializers_ = [];
+
+/**
+ * The function to handle unrecognized tags when parsing an XML document.
+ * @param {!Blockly.Workspace} workspace the Blockly workspace
+ * @param {!Element} xmlChild the XML element being parsed
+ * @param {!Blockly.Xml.ParserState} _state parse state information (not used
+ *     in this implementation, but may be used by projects that provide a
+ *     different behavior.
+ * @public
+ */
+Blockly.Xml.unrecognizedTagType = function(workspace, xmlChild, _state) {
+  if (xmlChild.nodeName.toLowerCase() != '#text' &&
```

Comment:
Whole lot of code repetition here.  Can we throw `xmlChild.nodeName.toLowerCase()` into a variable?